crates/modelardb_common/src/metadata/mod.rs

```diff
@@ -13,136 +13,1017 @@
+use tracing::warn;
 
 use crate::errors::ModelarDbError;
+use crate::metadata::compressed_file::CompressedFile;
+use crate::metadata::model_table_metadata::{GeneratedColumn, ModelTableMetadata};
+use crate::parser;
+use crate::types::{ErrorBound, Timestamp, UnivariateId, Value};
 
-/// The database providers that are currently supported by the metadata database.
-pub enum MetadataDatabaseType {
-    SQLite,
-    PostgreSQL,
+/// Name used for the file containing the SQLite database storing the metadata.
+const METADATA_DATABASE_NAME: &str = "metadata.sqlite3";
+
+/// The database providers that are currently supported by the metadata database. This trait is
+/// also used to handle small differences in the SQL syntax between database providers.
+pub trait MetadataDatabase {
+    /// Syntax added after a CREATE TABLE statement to specify that the table should use strict types.
+    fn strict(&self) -> &str;
+    /// Type used for binary columns in CREATE TABLE statements.
+    fn binary_type(&self) -> &str;
+}
+
+impl MetadataDatabase for Postgres {
+    fn strict(&self) -> &str {
+        ""
+    }
+
+    fn binary_type(&self) -> &str {
+        "BYTEA"
+    }
+}
+
+impl MetadataDatabase for Sqlite {
+    fn strict(&self) -> &str {
+        "STRICT"
+    }
+
+    fn binary_type(&self) -> &str {
+        "BLOB"
+    }
+}
+
+/// Stores the metadata required for reading from and writing to the tables and model tables.
+/// The data that needs to be persisted is stored in the metadata database.
+#[derive(Clone)]
+pub struct TableMetadataManager<DB: Database> {
+    /// The type of the database, used to handle small differences in SQL syntax between providers.
+    metadata_database_type: DB,
+    /// Pool of connections to the metadata database.
+    metadata_database_pool: Pool<DB>,
+    /// Cache of tag value hashes used to signify when to persist new unsaved tag combinations.
+    tag_value_hashes: DashMap<String, u64>,
 }
 
-/// If they do not already exist, create the tables in the metadata database used for table and
-/// model table metadata.
-/// * The table_metadata table contains the metadata for tables.
-/// * The model_table_metadata table contains the main metadata for model tables.
-/// * The model_table_hash_table_name contains a mapping from each tag hash to the name of the
-/// model table that contains the time series with that tag hash.
-/// * The model_table_field_columns table contains the name, index, error bound, and generation
-/// expression of the field columns in each model table.
-/// If the tables exist or were created, return [`Ok`], otherwise return [`Error`].
-pub async fn create_metadata_database_tables<'a, DB, E>(
-    metadata_database_pool: E,
-    database_type: MetadataDatabaseType,
-) -> Result<(), Error>
+impl<DB> TableMetadataManager<DB>
 where
-    DB: Database,
-    E: Executor<'a, Database = DB> + Copy,
+    DB: Database + MetadataDatabase,
```

Comment:
I have left it as it was for now since I think it is sufficient that each bound links directly to the SQLx documentation that explains what the bound does.